geode-common/src/main/java/org/apache/geode/internal/Retry.java

```diff
@@ -0,0 +1,101 @@
+ * copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package org.apache.geode.internal;
+
+import static java.util.concurrent.TimeUnit.NANOSECONDS;
+
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+
+import org.apache.geode.annotations.VisibleForTesting;
+
+/**
+ * Utility class for retrying operations.
+ */
+public class Retry {
+
+  interface Timer {
+    long nanoTime();
+
+    void sleep(long sleepTimeInNano) throws InterruptedException;
+  }
+
+  static class SteadyTimer implements Timer {
+    @Override
+    public long nanoTime() {
+      return System.nanoTime();
+    }
+
+    @Override
+    public void sleep(long sleepTimeInNano) throws InterruptedException {
+      long millis = NANOSECONDS.toMillis(sleepTimeInNano);
+      // avoid throwing IllegalArgumentException
+      if (millis > 0) {
+        Thread.sleep(millis);
+      }
+    }
+  }
+
+  private static final SteadyTimer steadyClock = new SteadyTimer();
+
+  /**
+   * Try the supplier function until the predicate is true or timeout occurs.
+   *
+   * @param timeout to retry for
+   * @param timeoutUnit the unit for timeout
+   * @param interval time between each try
+   * @param intervalUnit the unit for interval
+   * @param supplier to execute until predicate is true or times out
+   * @param predicate to test for retry
+   * @param <T> type of return value
+   * @return value from supplier after it passes predicate or times out.
+   */
+  public static <T> T tryFor(long timeout, TimeUnit timeoutUnit,
+      long interval, TimeUnit intervalUnit,
+      Supplier<T> supplier,
+      Predicate<T> predicate) throws TimeoutException, InterruptedException {
+    return tryFor(timeout, timeoutUnit, interval, intervalUnit, supplier, predicate, steadyClock);
+  }
+
+  @VisibleForTesting
+  static <T> T tryFor(long timeout, TimeUnit timeoutUnit,
+      long interval, TimeUnit intervalUnit,
+      Supplier<T> supplier,
+      Predicate<T> predicate,
+      Timer timer) throws TimeoutException, InterruptedException {
+    long until = timer.nanoTime() + NANOSECONDS.convert(timeout, timeoutUnit);
+    long intervalNano = NANOSECONDS.convert(interval, intervalUnit);
+
+    T value;
+    for (;;) {
+      value = supplier.get();
+      if (predicate.test(value)) {
+        return value;
+      } else {
+        // if there is still more time left after we sleep for interval period, then sleep and retry
+        // otherwise break out and throw TimeoutException
+        if ((timer.nanoTime() + intervalNano) < until) {
```

Comment:
There is no need to sleep for the full interval (`intervalNano`) if `until - timer.nanoTime()` is smaller.

I recommend (again), as have others, that the code sleep for the minimum time required. We understand that `Thread.sleep()` does not make real-time guarantees. Nevertheless, this code would be more robust and accurate if it was coded that way.

Something like this:
![image](https://user-images.githubusercontent.com/4002/99836122-778ff800-2b1a-11eb-96a0-da03974e9f94.png)

Reply:
@Bill per our offline meetings, can I merge this now?

Reply:
I think a user of the `Retry` class would (rightly) expect `tryFor()` to keep trying if `timeout` has not yet been reached.

Computing `sleepNanos` gives good results when `intervalUnit` is greater than (`timeoutUnit` - supplier time - predicate time). Without `sleepNanos`, that scenario results in _no retries at all_ which I think most users of the `Retry` class would find counterintuitive.

Failing to compute `sleepNanos` introduces an error. That error is directly proportional to `interval` and the relative error (error / `timeout`) grows as `interval` grows relative to `timeout`.

Computing `sleepNanos` is simple and cheap. It gives better accuracy and meets user expectations in all cases. Unless you foresee some problem computing `sleepNanos` I still recommend doing it.

Reply:
> I think a user of the `Retry` class would (rightly) expect `tryFor()` to keep trying if `timeout` has not yet been reached.
> 
> Computing `sleepNanos` gives good results when `intervalUnit` is greater than (`timeoutUnit` - supplier time - predicate time). Without `sleepNanos`, that scenario results in _no retries at all_ which I think most users of the `Retry` class would find counterintuitive.

So if say timeout is 5 second, supplier and predicate takes 3 seconds, and interval is 3 seconds, I won't expect there would be any retries at all. I don't think this is counterintuitive.


> Failing to compute `sleepNanos` introduces an error. That error is directly proportional to `interval` and the relative error (error / `timeout`) grows as `interval` grows relative to `timeout`.
> 
> Computing `sleepNanos` is simple and cheap. It gives better accuracy and meets user expectations in all cases. Unless you foresee some problem computing `sleepNanos` I still recommend doing it.

Reply:
A user of `tryFor()` does not know (usually cannot know) how much clock time their supplier and predicate will take. So they won't be able to do the reasoning you just did there. I think what they'll usually do is try to pick `timeout` much larger than the (supplier+predicate) time.

We can presume that they will usually try to set `timeout` long enough so that at least one retry (two tries) can happen. If we let`timeout` mean: keep trying until `timeout` has elapsed, then the "error" term I referred to in my previous comment is the difference between `timeout` and the actual time of the final attempt. As a user I'd expect `tryFor()` to make a reasonable effort to minimize that error.

Maybe it would help if, in my proposal,`interval` was interpreted as "typical sleep time". For all but the last iteration `interval` is the actual sleep time. Usually, in the last interval (under my proposal) `sleepNanos` will be less than `maximumSleepTime`.

Is there some reason why you object to (usually) sleeping a shorter time before the final try? Do you think that calling `min()` and capturing the result in a variable (`sleepNanos`) is overly burdensome? Do you see any value in making some effort to make the timing of the actual final attempt correspond to `timeout`?

Reply:
what's the point of sleeping if we know we are going to reach timeout after sleeping and throw a TimeoutException?

Reply:
After each `sleep()` (in my proposal) we try go back to the top of the loop and execute the supplier and the predicate again. We only `break` (and throw the exception) if `! (sleepNanos > 0)`)

Reply:
> A user of `tryFor()` does not know (usually cannot know) how much clock time their supplier and predicate will take. So they won't be able to do the reasoning you just did there. I think what they'll usually do is try to pick `timeout` much larger than the (supplier+predicate) time.
> 
> We can presume that they will usually try to set `timeout` long enough so that at least one retry (two tries) can happen. If we let`timeout` mean: keep trying until `timeout` has elapsed, then the "error" term I referred to in my previous comment is the difference between `timeout` and the actual time of the final attempt. As a user I'd expect `tryFor()` to make a reasonable effort to minimize that error.

No matter how long the `timeout` to be or how short the `interval` to be, a user can NOT "reasonably" assume two tries should happen at all. Even if I set timeout to be 100 seconds, and interval to be 1 second, there are plenty of chances that I could try only one time and the timeout period has already elapsed.

> 
> Maybe it would help if, in my proposal,`interval` was interpreted as "typical sleep time". For all but the last iteration `interval` is the actual sleep time. Usually, in the last interval (under my proposal) `sleepNanos` will be less than `maximumSleepTime`.
> 
> Is there some reason why you object to (usually) sleeping a shorter time before the final try? Do you think that calling `min()` and capturing the result in a variable (`sleepNanos`) is overly burdensome? Do you see any value in making some effort to make the timing of the actual final attempt correspond to `timeout`?

My objection is : 
1. there is no point in sleeping if we know we are going to timeout at the end of sleep.
2. It feels even more counter-intuitive that we know we've already reached the end of the timeout period and then still go ahead and try again. It's actually wrong.

Reply:
I don't understand (1). In my proposal we never time out at the end of a sleep. After a sleep we always try the supplier and predicate again. In cases where the predicate returns `false` forever, we would always sleep and then try the predicate and then `break` (and throw the exception.)

In cases where the predicate returns `false` forever, our final attempt is made as close to the timeout as possible (given the limitations of Java thread scheduling). This is exactly what a user expects.

Reply:
> In cases where the predicate returns `false` forever, our final attempt is made as close to the timeout as possible (given the limitations of Java thread scheduling). This is exactly what a user expects.

In your proposal, at your final attempt, you calculated what's the time left till timeout, then you sleep till timeout. Our differences lie at this point afterwards, I think that since we are already at timeout point we should throw exception, but you think we should try again. I don't think "trying again" is "exactly what a user expects", since I certainly didn't expect that.

Let's just list out the choices here and decide what's the "most" reasonable thing to do. Say our interval is 3 seconds, and there are only 2 seconds left after our last call to the supplier and the predicate is still false,  At this point, we can do either of the following:
1. throw timeout immediately (my implementation)
2. sleep for 2 seconds till timeout and throw timeout (since 2 is what's left till timeout)
3. sleep for 3 seconds and throw timeout (since 3 is what user has specified as interval)
4. sleep for 2 seconds till timeout and then try again (your suggestion)

In the above 4, my opinion is 4 is the least reasonable, and 1 has some optimization among the first 3.

Reply:
The usage of `Retry` (as it stands) in this PR, by `tryToResolveLocalHost()` is fine:

`tryFor(60, SECONDS, 1, SECONDS…`

The difference between (1) and (4) in this case is at most:

(1) last retry at 59 seconds
(4) last retry at 60 seconds

If the purpose of this PR were to make the best possible `Retry` utility, I might be inclined to continue the deliberation. But since the purpose of this PR is to retry resolving the localhost address, and the PR does that, my work is done here.

In the interest of forward progress, I withdraw my change request on this PR.